runtime/src/snapshot_utils/snapshot_storage_rebuilder.rs

```diff
@@ -0,0 +1,363 @@
+        let snapshot_storage_lengths = Self::process_snapshot_file(snapshot_file_path).unwrap();
+        Self::spawn_rebuilder_threads(
+            file_receiver,
+            num_threads,
+            next_append_vec_id,
+            snapshot_storage_lengths,
+            append_vec_files,
+        )
+    }
+
+    /// Create the SnapshotStorageRebuilder for storing state during rebuilding
+    ///     - pre-allocates data for storage paths
+    fn new(
+        file_receiver: Receiver<PathBuf>,
+        num_threads: usize,
+        next_append_vec_id: Arc<AtomicAppendVecId>,
+        snapshot_storage_lengths: HashMap<Slot, HashMap<usize, usize>>,
+    ) -> Self {
+        let storage = DashMap::with_capacity(snapshot_storage_lengths.len());
+        let storage_paths: DashMap<_, _> = snapshot_storage_lengths
+            .iter()
+            .map(|(slot, storage_lengths)| {
+                (*slot, Mutex::new(Vec::with_capacity(storage_lengths.len())))
+            })
+            .collect();
+        Self {
+            file_receiver,
+            num_threads,
+            snapshot_storage_lengths,
+            storage_paths,
+            storage,
+            next_append_vec_id,
+            processed_slot_count: AtomicUsize::new(0),
+            num_collisions: AtomicUsize::new(0),
+        }
+    }
+
+    /// Waits for snapshot file
+    /// Due to parallel unpacking, we may receive some append_vec files before the snapshot file
+    /// This function will push append_vec files into a buffer until we receive the snapshot file
+    fn get_snapshot_file(file_receiver: &Receiver<PathBuf>) -> (PathBuf, Vec<PathBuf>) {
+        let mut append_vec_files = Vec::with_capacity(1024);
+        let snapshot_file_path = loop {
+            if let Ok(path) = file_receiver.recv() {
+                let filename = path.file_name().unwrap().to_str().unwrap();
+                match get_snapshot_file_kind(filename) {
+                    Some(SnapshotFileKind::SnapshotFile) => {
+                        break path;
+                    }
+                    Some(SnapshotFileKind::StorageFile) => {
+                        append_vec_files.push(path);
+                    }
+                    None => {} // do nothing for other kinds of files
+                }
+            } else {
+                panic!("did not receive snapshot file from unpacking threads");
+            }
+        };
+
+        (snapshot_file_path, append_vec_files)
+    }
+
+    /// Process the snapshot file to get the size of each snapshot storage file
+    fn process_snapshot_file(
+        snapshot_file_path: PathBuf,
+    ) -> Result<HashMap<Slot, HashMap<usize, usize>>, bincode::Error> {
+        let snapshot_file = File::open(snapshot_file_path).unwrap();
+        let mut snapshot_stream = BufReader::new(snapshot_file);
+        let (_bank_fields, accounts_fields) =
+            serde_snapshot::fields_from_stream(SerdeStyle::Newer, &mut snapshot_stream)?;
```

Comment:
I'm wondering if we will need to handle `Newer` here more generically or with a parameter. I think for now we are ok, but this may be a hiccup when we bump the persistence version.

Reply:
Hmm that's a good point, I didn't think this through with there currently only being the single version. We probably need to actually check the version to get SerdeStyle to do this properly though.

Currently the version file is serialized after accounts, so we'd have to move that to before like I did with the snapshots(bank) file already.

Reply:
#27192

Reply:
for now, this is probably ok. I imagine we'll fail when someone tries to bump the version.

Reply:
should probably create an issue to deal with this.

Reply:
#27346